sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Avro/SchemaRegistryAvroObjectSerializer.cs

```diff
@@ -0,0 +1,241 @@
+            stream.Write(Encoding.UTF8.GetBytes(schemaId), 0, 32);
+            writer.Write(value, binaryEncoder);
+            binaryEncoder.Flush();
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask SerializeAsync(Stream stream, object value, Type inputType, CancellationToken cancellationToken)
+        {
+            Argument.AssertNotNull(stream, nameof(stream));
+            Argument.AssertNotNull(value, nameof(value));
+            Argument.AssertNotNull(inputType, nameof(inputType));
+
+            var supportedType = GetSupportedTypeOrThrow(inputType);
+            var writer = GetWriterAndSchema(value, supportedType, out var schema);
+            var schemaId = await GetSchemaIdAsync(schema, cancellationToken).ConfigureAwait(false);
+
+            var binaryEncoder = new BinaryEncoder(stream);
+            await stream.WriteAsync(s_emptyRecordFormatIndicator, 0, 4, cancellationToken).ConfigureAwait(false);
+            await stream.WriteAsync(Encoding.UTF8.GetBytes(schemaId), 0, 32, cancellationToken).ConfigureAwait(false);
+            writer.Write(value, binaryEncoder);
+            binaryEncoder.Flush();
+        }
+
+        private Schema GetSchemaById(string schemaId, CancellationToken cancellationToken)
+        {
+            if (_cachedSchemas.TryGetValue(schemaId, out var cachedSchema))
+            {
+                return cachedSchema;
+            }
+
+            var schemaContent = _client.GetSchema(schemaId, cancellationToken).Value.Content;
+            var schema = Schema.Parse(schemaContent);
+            _cachedSchemas.Add(schemaId, schema);
+            return schema;
+        }
+
+        private async Task<Schema> GetSchemaByIdAsync(string schemaId, CancellationToken cancellationToken)
+        {
+            if (_cachedSchemas.TryGetValue(schemaId, out var cachedSchema))
+            {
+                return cachedSchema;
+            }
+
+            var schemaContent = (await _client.GetSchemaAsync(schemaId, cancellationToken).ConfigureAwait(false)).Value.Content;
+            var schema = Schema.Parse(schemaContent);
+            _cachedSchemas.Add(schemaId, schema);
+            return schema;
+        }
+
+        private static DatumReader<object> GetReader(Schema schema, SupportedType supportedType)
+        {
+            switch (supportedType)
+            {
+                case SupportedType.SpecificRecord:
+                    return new SpecificDatumReader<object>(schema, schema);
+                case SupportedType.GenericRecord:
+                    return new GenericDatumReader<object>(schema, schema);
+                default:
+                    throw new ArgumentException($"Invalid supported type value: {supportedType}");
+            }
+        }
+
+        private static ReadOnlyMemory<byte> CopyToReadOnlyMemory(Stream stream)
+        {
+            using var tempMemoryStream = new MemoryStream();
+            stream.Position = 0;
```

Comment:
Why do we set to 0? What if user doesn't want to deserialize from start of stream?

Reply:
Ah! That makes a ton of sense. I'll fix that. It should be the onus of the user passing in the stream to set the position they want to use.

Reply:
Fixed.